assets/js/googlesitekit/modules/datastore/modules.js

```diff
@@ -149,6 +178,51 @@ const baseActions = {
 
 		return { response, error };
 	},
+
+	/**
+	 * Registers a module.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {string}    slug                         Module slug.
+	 * @param {Object}    [settings]                   Optional. Module settings.
+	 * @param {string}    [settings.name]              Optional. Module name. Default is the slug.
+	 * @param {string}    [settings.description]       Optional. Module description. Default empty string.
+	 * @param {string}    [settings.icon]              Optional. Module icon. Default empty string.
+	 * @param {number}    [settings.order]             Optional. Numeric indicator for module order. Default 10.
+	 * @param {string}    [settings.homepage]          Optional. Module homepage URL. Default empty string.
+	 * @param {WPElement} [settings.settingsComponent] React component to render the settings panel. Default is the DefaultModuleSettings component.
+	 * @return {Object} Redux-style action.
+	 */
+	*registerModule( slug, { settingsComponent = DefaultModuleSettings, ...settings } = {} ) {
+		invariant( slug, 'module slug is required' );
+
```

Comment:
Something that just came to mind: We'll need to make sure server-side modules have been loaded (via `getModules` and its resolver) before running the logic below. Especially since these `registerModule` calls will likely happen early during app lifecycle.

See my other comment, we may need to explicitly wait for that? cc @tofumatt

Reply:
Based on https://github.com/google/site-kit-wp/pull/1671#pullrequestreview-453736379, we'll probably need some internal `waitForModules` action with a control (similar like we have e.g. `waitForExistingTag`).